wasmtime-environ/src/cache.rs

```diff
@@ -0,0 +1,300 @@
+                #[cfg(windows)]
+                let mut long_path = OsString::from("\\\\?\\");
+                #[cfg(windows)]
+                let cache_dir = {
+                    if cache_dir.starts_with("\\\\?\\") {
+                        cache_dir
+                    }
+                    else {
+                        long_path.push(cache_dir.as_os_str());
+                        Path::new(&long_path)
+                    }
+                };
+                match fs::create_dir_all(cache_dir) {
+                    Ok(()) => (),
+                    Err(err) => warn!("Unable to create cache directory, failed with: {}", err),
+                };
+                Some(cache_dir.to_path_buf())
+            }
+            None => {
+                warn!("Unable to find cache directory");
+                None
+            }
+        };
+}
+
+pub struct ModuleCacheEntry {
+    mod_cache_path: Option<PathBuf>,
+}
+
+#[derive(serde::Serialize, serde::Deserialize)]
+pub struct ModuleCacheData {
+    compilation: Compilation,
+    relocations: Relocations,
+    address_transforms: ModuleAddressMap,
+}
+
+type ModuleCacheDataTupleType = (Compilation, Relocations, ModuleAddressMap);
+
+impl ModuleCacheEntry {
+    pub fn new(module: &Module, _isa: &dyn isa::TargetIsa, _generate_debug_info: bool) -> Self {
+        // TODO: cache directory hierarchy with isa name, compiler name & git revision, and files with flag if debug symbols are available
+        let option_hash = module.hash;
+
+        let mod_cache_path = CACHE_DIR.clone().and_then(|p| {
+            option_hash.map(|hash| {
+                p.join(format!(
+                    "mod-{}",
+                    base64::encode_config(&hash, base64::URL_SAFE_NO_PAD) // standard encoding uses '/' which can't be used for filename
+                ))
+            })
+        });
+
+        ModuleCacheEntry { mod_cache_path }
+    }
+
+    pub fn get_data(&self) -> Option<ModuleCacheData> {
+        if let Some(p) = &self.mod_cache_path {
+            match fs::read(p) {
+                Ok(cache_bytes) => match bincode::deserialize(&cache_bytes[..]) {
+                    Ok(data) => Some(data),
+                    Err(err) => {
+                        warn!("Failed to deserialize cached code: {}", err);
+                        None
+                    }
+                },
+                Err(_) => None,
+            }
+        } else {
+            None
+        }
+    }
+
+    pub fn update_data(&self, data: &ModuleCacheData) {
+        if let Some(p) = &self.mod_cache_path {
+            let cache_buf = match bincode::serialize(&data) {
+                Ok(data) => data,
+                Err(err) => {
+                    warn!("Failed to serialize cached code: {}", err);
+                    return;
+                }
+            };
+            match fs::write(p, &cache_buf) {
+                Ok(()) => (),
+                Err(err) => warn!(
```

Comment:
If I understand correctly, if the write fails, `fs::write` may leave a partially-written file behind. We should do an `fs::remove_file` on the path in the error case.

Reply:
Do you mean calling it anyway? I'm not sure how I should detect this situation from the [Error](https://doc.rust-lang.org/std/io/struct.Error.html). It's good idea to cleanup, so I'll do it, however current code is still correct: next time the file would be read, not parsed correctly, and overwritten with new cached data.